File: backend/schemas/module2.py
"""M2 — Overnight Gap Analysis schemas."""
from __future__ import annotations

from typing import Any
from pydantic import BaseModel, Field


class Module2Request(BaseModel):
    instrument: str = Field("NIFTY", pattern="^(NIFTY|BANKNIFTY)$")
    from_date: str          # YYYY-MM-DD
    to_date: str            # YYYY-MM-DD
    pts_to_analyse: int = Field(100, ge=25, le=500)
    option_range: int = Field(3, ge=1, le=10)
    expiry_count: int = Field(2, ge=1, le=5)


class SessionRow(BaseModel):
    date: str               # entry date (Day D)
    entry_time: str         # 15:15
    exit_time: str          # 09:25 next day
    gap_pts: float          # signed T1-T0
    gap_direction: str      # GapUp | GapDown
    vix: float | None
    start_price: float
    end_price: float
    columns: dict[str, Any]


class Module2Result(BaseModel):
    instrument: str
    from_date: str
    to_date: str
    pts_to_analyse: int
    option_range: int
    expiry_count: int
    sessions: list[SessionRow]
    column_meta: dict[str, Any]
    run_id: str

File: frontend/src/api/module2.ts
import { apiClient } from './client';

export interface Module2Request {
  instrument: 'NIFTY' | 'BANKNIFTY';
  from_date: string;
  to_date: string;
  pts_to_analyse: number;
  option_range: number;
  expiry_count: number;
}

export interface SessionRow {
  date: string;
  entry_time: string;
  exit_time: string;
  gap_pts: number;
  gap_direction: 'GapUp' | 'GapDown';
  vix: number | null;
  start_price: number;
  end_price: number;
  columns: Record<string, number | null>;
}

export interface ColumnMeta {
  label: string;
  group: 'B' | 'C' | 'E';
  exp_n?: number;
  strike_label?: string;
  side?: string;
  pct?: number;
}

export interface Module2Result {
  instrument: string;
  from_date: string;
  to_date: string;
  pts_to_analyse: number;
  option_range: number;
  expiry_count: number;
  sessions: SessionRow[];
  column_meta: Record<string, ColumnMeta>;
  run_id: string;
}

export async function runModule2(req: Module2Request): Promise<Module2Result> {
  const { data } = await apiClient.post<Module2Result>('/module2/run', req);
  return data;
}

export async function getModule2Result(runId: string): Promise<Module2Result> {
  const { data } = await apiClient.get<Module2Result>(`/module2/results/${runId}`);
  return data;
}
